[Python][Flaky Test] test_update_connection_password - TimeoutError in cluster mode

### Test Name

test_update_connection_password

### Test Location

tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password
tests/sync_tests/test_sync_auth.py::TestSyncAuthCommands::test_sync_update_connection_password

### Failure Permlink

https://github.com/valkey-io/valkey-glide/actions/runs/28420121702/job/84211287744#logs

### Frequency

Observed in scheduled CI run on 2026-06-30. Failed in all 84 Python test matrix jobs (cluster mode True only).

### Steps to Reproduce

1. Run the FMT Python workflow on schedule
2. Observe failure in Python Tests jobs with cluster mode enabled

### System Information

GitHub Actions (ubuntu-24.04, x86_64 and aarch64)

### Language and Version

Python 3.9 - 3.14

### Engine Version

All engine versions (6.2, 7.0, 7.2, 8.0, 8.1, 9.0)

### Logs

```
FAILED tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password[asyncio-ProtocolVersion.RESP2-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password[asyncio-ProtocolVersion.RESP3-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password[trio-ProtocolVersion.RESP2-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password[trio-ProtocolVersion.RESP3-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_auth.py::TestSyncAuthCommands::test_sync_update_connection_password[ProtocolVersion.RESP2-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_auth.py::TestSyncAuthCommands::test_sync_update_connection_password[ProtocolVersion.RESP3-True] - glide_shared.exceptions.TimeoutError: timed out

= 7 failed, 5537 passed, 2214 skipped, 2229 deselected, 31 warnings in 2990.30s (0:49:50) =

Stderr:
Error executing command: Cluster: Failed to receive command response from internal sender. RecvError(())
Error executing command: Failed to receive a response due to a fatal error - FatalReceiveError: channel closed
```

### Screenshots

_No response_

### Glide Version

main

### Expected Behavior

Test should pass reliably.

### Actual Behavior

The test_update_connection_password test times out consistently in cluster mode (True parameter) across all Python versions, engine versions, and async backends (asyncio, trio). The timeout occurs during connection password update operations. The error indicates the cluster connection channel is closed or fails to receive responses, suggesting a race condition or timing issue in the password update flow when used with cluster connections.

### Possible Fixes

The test only fails with cluster_mode=True. The timeout and RecvError/FatalReceiveError suggest the connection is being dropped during the password update. Possible fixes:
1. Increase timeout for password update operations in cluster mode
2. Add retry logic for connection re-establishment after password update
3. Investigate if CLUSTER AUTH propagation timing causes connection drops

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -30,6 +30,7 @@ from tests.utils.utils import (
     delete_acl_username_and_password,
     kill_connections,
     set_new_acl_username_with_password,
+    wait_for,
 )
 
 
@@ -103,20 +104,36 @@ class TestAuthCommands:
         assert value == b"test_value"
         await config_set_new_password(glide_client, NEW_PASSWORD)
         await kill_connections(management_client)
-        # Add a short delay to allow the server to apply the new password
-        # without this delay, command may or may not time out while the client reconnect
-        # ending up with a flaky test
-        await anyio.sleep(2)
-        # Verify that the client is able to reconnect with the new password,
-        value = await glide_client.get("test_key")
-        assert value == b"test_value"
-        await kill_connections(management_client)
-        await anyio.sleep(2)
-        # Verify that the client is able to immediateAuth with the new password after client is killed
-        result = await glide_client.update_connection_password(
-            NEW_PASSWORD, immediate_auth=True
+        # Wait for the client to reconnect with the new password using retry
+        # instead of a fixed sleep, which is unreliable in cluster mode under CI load
+        async def _check_reconnected_after_first_kill():
+            try:
+                val = await glide_client.get("test_key")
+                return val == b"test_value"
+            except Exception:
+                return False
+
+        await wait_for(
+            _check_reconnected_after_first_kill,
+            "Client failed to reconnect with new password after first kill",
+            timeout=15,
         )
-        assert result == OK
+        await kill_connections(management_client)
+        # Wait for reconnection again before attempting immediate auth
+        async def _check_reconnected_after_second_kill():
+            try:
+                result = await glide_client.update_connection_password(
+                    NEW_PASSWORD, immediate_auth=True
+                )
+                return result == OK
+            except Exception:
+                return False
+
+        await wait_for(
+            _check_reconnected_after_second_kill,
+            "Client failed to reconnect after second kill for immediate auth",
+            timeout=15,
+        )
         # Verify that the client is still authenticated
         assert await glide_client.set("test_key", "test_value") == OK
 

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -117,7 +117,7 @@ class TestAuthCommands:
         await wait_for(
             _check_reconnected_after_first_kill,
             "Client failed to reconnect with new password after first kill",
-            timeout=15,
+            timeout=25,
         )
         await kill_connections(management_client)
 
@@ -134,7 +134,7 @@ class TestAuthCommands:
         await wait_for(
             _check_reconnected_after_second_kill,
             "Client failed to reconnect after second kill for immediate auth",
-            timeout=15,
+            timeout=25,
         )
         # Verify that the client is still authenticated
         assert await glide_client.set("test_key", "test_value") == OK

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -104,6 +104,7 @@ class TestAuthCommands:
         assert value == b"test_value"
         await config_set_new_password(glide_client, NEW_PASSWORD)
         await kill_connections(management_client)
+
         # Wait for the client to reconnect with the new password using retry
         # instead of a fixed sleep, which is unreliable in cluster mode under CI load
         async def _check_reconnected_after_first_kill():
@@ -119,6 +120,7 @@ class TestAuthCommands:
             timeout=15,
         )
         await kill_connections(management_client)
+
         # Wait for reconnection again before attempting immediate auth
         async def _check_reconnected_after_second_kill():
             try:

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -31,6 +31,7 @@ from tests.utils.utils import (
     delete_acl_username_and_password,
     kill_connections,
     set_new_acl_username_with_password,
+    sync_wait_for,
 )
 
 
@@ -103,20 +104,34 @@ class TestSyncAuthCommands:
         assert value == b"test_value"
         config_set_new_password(glide_sync_client, NEW_PASSWORD)
         kill_connections(management_sync_client)
-        # Add a short delay to allow the server to apply the new password
-        # without this delay, command may or may not time out while the client reconnect
-        # ending up with a flaky test
-        time.sleep(2)
-        # Verify that the client is able to reconnect with the new password,
-        value = glide_sync_client.get("test_key")
-        assert value == b"test_value"
-        kill_connections(management_sync_client)
-        time.sleep(2)
-        # Verify that the client is able to immediateAuth with the new password after client is killed
-        result = glide_sync_client.update_connection_password(
-            NEW_PASSWORD, immediate_auth=True
+        # Wait for the client to reconnect with the new password using retry
+        # instead of a fixed sleep, which is unreliable in cluster mode under CI load
+        def _check_reconnected_after_first_kill():
+            try:
+                val = glide_sync_client.get("test_key")
+                return val == b"test_value"
+            except Exception:
+                return False
+
+        sync_wait_for(
+            _check_reconnected_after_first_kill,
+            "Client failed to reconnect with new password after first kill",
         )
-        assert result == OK
+        kill_connections(management_sync_client)
+        # Wait for reconnection again before attempting immediate auth
+        def _check_reconnected_after_second_kill():
+            try:
+                result = glide_sync_client.update_connection_password(
+                    NEW_PASSWORD, immediate_auth=True
+                )
+                return result == OK
+            except Exception:
+                return False
+
+        sync_wait_for(
+            _check_reconnected_after_second_kill,
+            "Client failed to reconnect after second kill for immediate auth",
+        )
         # Verify that the client is still authenticated
         assert glide_sync_client.set("test_key", "test_value") == OK
 

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -117,6 +117,7 @@ class TestSyncAuthCommands:
         sync_wait_for(
             _check_reconnected_after_first_kill,
             "Client failed to reconnect with new password after first kill",
+            timeout=25,
         )
         kill_connections(management_sync_client)
 
@@ -133,6 +134,7 @@ class TestSyncAuthCommands:
         sync_wait_for(
             _check_reconnected_after_second_kill,
             "Client failed to reconnect after second kill for immediate auth",
+            timeout=25,
         )
         # Verify that the client is still authenticated
         assert glide_sync_client.set("test_key", "test_value") == OK

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -104,6 +104,7 @@ class TestSyncAuthCommands:
         assert value == b"test_value"
         config_set_new_password(glide_sync_client, NEW_PASSWORD)
         kill_connections(management_sync_client)
+
         # Wait for the client to reconnect with the new password using retry
         # instead of a fixed sleep, which is unreliable in cluster mode under CI load
         def _check_reconnected_after_first_kill():
@@ -118,6 +119,7 @@ class TestSyncAuthCommands:
             "Client failed to reconnect with new password after first kill",
         )
         kill_connections(management_sync_client)
+
         # Wait for reconnection again before attempting immediate auth
         def _check_reconnected_after_second_kill():
             try:

--- a/python/tests/utils/utils.py
+++ b/python/tests/utils/utils.py
@@ -322,19 +322,21 @@ async def wait_for(
 def sync_wait_for(
     condition: Callable[[], bool],
     failure: str,
+    timeout: float = _WAIT_FOR_TIMEOUT_SEC,
 ) -> None:
     """Waits until a condition is met.
 
     Args:
         condition: Callable that returns True when the condition is met.
         failure: Error message raised if the condition is not met within timeout.
+        timeout: Maximum time to wait for the condition to be met, in seconds.
 
     Raises:
         TimeoutError: If the condition is not met within the timeout.
     """
     import time as _time
 
-    deadline = _time.monotonic() + _WAIT_FOR_TIMEOUT_SEC
+    deadline = _time.monotonic() + timeout
 
     while _time.monotonic() < deadline:
         if condition():

